docs_src/design/adr/014-Secret-Provider-For-All.md

```diff
@@ -0,0 +1,398 @@
+
+### Only Exclusive Secret Stores
+
+To simplify the `SecretProvider` abstraction, we need to reduce to using only exclusive `SecretStores`. This allows all the APIs to deal with a single `SecretClient`, rather than the split up way we currently have in Application Services. This requires that the current Application Service shared secrets (database credentials) must be copied into each Application Service's exclusive `SecretStore` when it is created.
+
+The challenge is how do we seed known secrets for unknown services when they become known.  As described above in the [Known and Unknown Services](#known-and-unknown-services) section above,  services currently identify themselves for exclusive `SecretStore` creation via the `ADD_SECRETSTORE_TOKENS` environment variable on security-secretstore-setup. This environment variable simply takes a comma separated list of service names.
+
+```yaml
+ADD_SECRETSTORE_TOKENS: "<service-name1>,<service-name2>"
+```
+
+If we expanded this to add an optional list of known secret identifiers for each service, i.e.  `appservice/redisdb`, the exclusive store could also be seeded with a copy of known shared secrets. In this case the Redis database credentials for the Application Services' shared database. The environment variable name will change to `ADD_SECRETSTORE` now that it is more than just tokens.
+
+```yaml
+ADD_SECRETSTORE: "app-service-xyz[appservice/redisdb]"
+```
+
+> *Note: The secret identifier here is the short path to the secret in the existing **appservice**  `SecretStore`. In the above example this expands to the full path of `/secret/edgex/appservice/redisdb`*
+
+The above example results in the Redis credentials being copied into app-service-xyz's `SecretStore` at `/secret/edgex/app-service-xyz/redis`.
+
+Similar approach could be taken for Message Bus credentials where a common `SecretStore` is created with the Message Bus credentials saved. The services request the credentials are copied into their exclusive `SecretStore` using `common/messagebus` as the secret identifier.
+
+Full specification for the environment variable's value is a comma separated list of service entries defined as:
+
+```
+<service-name>[optional list of known secret IDs sperated by ;]>
+```
+
+Example with one service specifying IDs for know secrets and one without known secrets 
+
+```yaml
+ADD_SECRETSTORE: "appservice-xyz[appservice/redisdb; common/messagebus], appservice-http-export"
+```
+
+When the `ADD_SECRETSTORE` environment variable is processed to create these `SecretStores`, it will copy the specified saved secrets from the initial `SecretStore` into the service's `SecretStore`. This all depends on the completion of database or other credential bootstrapping and the secrets having been stored prior to the environment variable being processed. security-secretstore-setup will need to be refactored to ensure this sequencing.
```

Comment:
Q:  When to ensure this sequencing, will we introduce the circular dependency in startup?  Since the `security-secretstore-setup` need to wait for the `redis's dbcrendtials` secretStore to be seeded initially but at the same time `redis's` db container is also waiting for the `security-secretstore-setup` to signal done to proceed, right?  It looks like there is a race condition between these two containers: a catch-22 scenario to me OR at least this is not crystal clear to me on how this can happen today yet.

Reply:
A.  Today, the code that processes ADD_SECRETSTORE runs before credential bootstrapping.  This needs to be changed for this to work.

A: Security secretstore setup today writes a flag SECRETSTORE_SETUP_DONE_FLAG that is watched by consul to signal that all of the secretstore bootstrapping is done.  It doesn't appear that redis-bootstrap blocks on this, or I am not seeing it.  Potential bug.

A: SECRETSTORE_SETUP_DONE_FLAG is going to be incorporated into the ready_to_run socket in the bootstrapping ADR so should be no concern there.

A: Potential issues with redis not remembering its provisioned password.